demo.py

```diff
@@ -0,0 +1,146 @@
+class Resource(models.Model):
+    client = ForeignKey(Client)
+    owned_by = ForgeignKey(Resource, null=True)
+    name = CharField(blank=False, max_length=128)
+    configuration = JSONField(null=False)
+    types = ResourceTypeRegistry()
+    owned_resources = ResourceManger()
+
+    @property
+    def configuration_schema(self):
+        """ Define the options used to configure this resource. """
+        return {}  # No configurable options by default
+
+    @property
+    def state(self):
+        """
+        The current state of this resource.
+        The return value is always a subclass of ResourceState, and typically will be a subclass of:
+          NeedsRefreshState, ApplyingConfigurationState, NeedsInputState, ReadyState, ErrorState, or DestroyingState
+        Each state supports common methods like get_status_message(), get_status_color(), allow_configuration_changes()
+        """
+        return self._state
+
+    def build_form():
+        """ Convert self.configuration_schema into a Django form """
+        pass
```

Comment:
See https://github.com/open-craft/opencraft/pull/28/files#r44956742 - if subclasses don't have "real" fields and instead store their fields in `configuration`, that's why. Plus this is more flexible without being *that* complicated - we are still using Django forms, just not Django ModelForms.

But keep in mind, this is just an implementation detail and not a key part of the overall concept here.